### PR TITLE
indexes_of function

### LINKED ISSUE #22: New function: F\indexes_of()
Returns a list of array indexes, either matching the predicate or strictly equal to the the passed value.

```php
array Functional\indexes_of(Traversable|array $collection, mixed|callable $predicate)
```

### PR DESCRIPTION
**Pull request by mschwager**
Fixes #22

### CODE REVIEW

File: src/Functional/IndexesOf.php [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+function indexes_of($collection, $value)
+{
+    InvalidArgumentException::assertCollection($collection, __FUNCTION__, 1);
+
+    $result = array();
+
+    if (is_callable($value)) {
+        foreach ($collection as $index => $element) {
+            if ($element === $value($element, $index, $collection)) {
+                array_push($result, $index);
```
Comment (lstrojny) [inline src/Functional/IndexesOf.php]: $result[] = …;

File: src/Functional/IndexesOf.php [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+            }
+        }
+    } else {
+        foreach ($collection as $index => $element) {
+            if ($element === $value) {
+                array_push($result, $index);
```
Comment (lstrojny) [inline src/Functional/IndexesOf.php]: $result[] = …;

File: src/Functional/IndexesOf.php [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+ *
+ * @param Traversable|array $collection
+ * @param mixed|callable $value
+ * @return array
+ */
+function indexes_of($collection, $value)
+{
+    InvalidArgumentException::assertCollection($collection, __FUNCTION__, 1);
+
+    $result = array();
```
Comment (lstrojny) [inline src/Functional/IndexesOf.php]: We can use the short array syntax now

File: tests/unit/Functional/IndexesOf.php [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+        $this->assertSame(array(3), indexes_of($this->iterator, 'value2'));
+        $this->assertSame(array('k1', 'k3'), indexes_of($this->hash, 'val1'));
+        $this->assertSame(array('k1', 'k3'), indexes_of($this->hashIterator, 'val1'));
+        $this->assertSame(array('k2'), indexes_of($this->hash, 'val2'));
+        $this->assertSame(array('k2'), indexes_of($this->hashIterator, 'val2'));
+        $this->assertSame(array('k4'), indexes_of($this->hash, 'val3'));
+        $this->assertSame(array('k4'), indexes_of($this->hashIterator, 'val3'));
+    }
+
+    function testIfValueCouldNotBeFoundFalseIsReturned()
```
Comment (lstrojny) [inline tests/unit/Functional/IndexesOf.php]: It’s not true that we return false, is it?

File: tests/unit/Functional/IndexesOf.php [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+    {
+        $this->assertEmpty(indexes_of($this->array, 'invalidValue'));
```
Comment (lstrojny) [inline tests/unit/Functional/IndexesOf.php]: A more explicit assertion would be `$this->assertSame([], index_of(…))`

**Comment by mschwager** [comment]
Also added a PHPUnit fix.

```bash
$ phpunit -c phpunit.xml.dist 
PHPUnit 4.7.2 by Sebastian Bergmann and contributors.

Runtime:	PHP 5.6.10-1+deb.sury.org~utopic+1 with Xdebug 2.3.2
Configuration:	/home/matt/documents/functional-php/phpunit.xml.dist

...............................................................  63 / 300 ( 21%)
............................................................... 126 / 300 ( 42%)
............................................................... 189 / 300 ( 63%)
............................................................... 252 / 300 ( 84%)
................................................

Time: 5.46 seconds, Memory: 28.00Mb

OK (300 tests, 1003 assertions)

Generating code coverage report in Clover XML format ... done

Generating code coverage report in HTML format ... done
```

**Comment by lstrojny** [comment]
Looks good! Can you add docs and squash the commits and then we are ready to merge.

**Comment by lstrojny** [comment]
Thank you!

**Pull request merged**